juno_magic/extensions/wamp.py

```diff
@@ -182,34 +184,89 @@ def onJoin(self, details):
             yield self.register(self)
             log.msg("[onJoin] ...done.")
             log.msg("[onJoin] Checking in with Magics class")
-            magics_instance.set_connection(self)
+            yield magics_instance.set_connection(self)
             log.msg("[onJoin] ...done.")
             print("Successfully connected to {}".format(magics_instance._router_url))
             if magics_instance._kernel_prefix:
                 print("Attempting to reconnect to {}".format(magics_instance._kernel_prefix))
                 yield self.set_prefix(magics_instance._kernel_prefix)
                 print("Reconnected to kernel prefix {}".format(magics_instance._kernel_prefix))
-                if not magics_instance._heartbeat.running:
-                    magics_instance._heartbeat.start(magics_instance._hb_interval, now=False)
+            if not magics_instance._heartbeat.running:
+                magics_instance._heartbeat.start(magics_instance._hb_interval, now=False)
             returnValue(None)
 
         @inlineCallbacks
         def onLeave(self, details):
             log.msg("[WampConnectionComponent] onLeave()")
             log.msg("details: {}".format(str(details)))
-            yield super(self.__class__, self).onLeave(details)
-            magics_instance.set_connection(None)
-            log.msg("set magics connection to None")
-            returnValue(None)
+            magics_instance._wamp_err_handler(details)
+            res =  yield super(WampConnectionComponent, self).onLeave(details)
+            returnValue(res)
 
         def onDisconnect(self):
-            # onDisconnect we should just set the connection to None so that we know to reconnect
-            # next time connect is called
-            #magics_instance.set_connection(None)
             pass
 
     return WampConnectionComponent
 
+
+class WampErrorDispatcher(Component):
+    exception = None
+    def __init__(self, magic, module='timbr.machine', **kwargs):
+        self.magic = magic
+        self._module = module
+        super(WampErrorDispatcher, self).__init__(target_name=module,  **kwargs)
+
+    def __call__(self, failure):
```

Comment:
This is cool, so any time we just call this class it will emit a comm message to the front end